optuna/storages/base.py

```diff
@@ -24,162 +24,848 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
+            template_trial:
+                Fronzen trial with default user-attributes, system-attributes,
+                intermediate-values, and a state.
+
+        Returns:
+            Trial ID of the created trial.
+
+        Raises (current implementation):
+            in-memory:
+                ValueError:
+                    If no study with the matching `study_id` exists.
+            rdb:
+                sqlalchemy.exc.IntegrityError:
+                    If no study with the matching `study_id` exists.
+            redis:
+                AssertionError:
+                    If no study with the matching `study_id` exists.
+
+        Raises (proposal):
+            KeyError:
+                If no study with the matching `study_id` exists.
+        """
 
         raise NotImplementedError
 
     @abc.abstractmethod
-    def set_trial_state(self, trial_id, state):
-        # type: (int, TrialState) -> bool
+    def set_trial_state(self, trial_id: int, state: TrialState) -> bool:
+
+        """Update a state of a specified trial.
+
+        Args:
+            trial_id:
+                Trial ID of a trial to update its state.
+            state:
+                New state of the trial.
+
+        Returns:
+
+        Raise (current implementation):
+            in-memory:
+                IndexError:
+                    If no trial with the matching `trial_id` exists.
+                RuntimeError:
+                    If the trial is already finished.
+            rdb:
+                ValueError:
+                    If no trial with the matching `trial_id` exists.
+                RuntimeError:
+                    If the trial is already finished.
+            redis:
+                AssertionError:
+                    If no trial with the matching `trial_id` exists.
+                RuntimeError:
+                    If the trial is already finished.
+
+        Raises (proposal):
+            KeyError:
+                If no trial with the matching `trial_id` exists.
+            RuntimeError:
+                If the trial is already finished.
+        """
 
         raise NotImplementedError
 
     @abc.abstractmethod
-    def set_trial_param(self, trial_id, param_name, param_value_internal, distribution):
-        # type: (int, str, float, distributions.BaseDistribution) -> bool
+    def set_trial_param(self, trial_id: int, param_name: str, param_value_internal: float, distribution: distributions.BaseDistribution) -> bool:
+
+        """Add a parameter to a specified trial.
+
+        Args:
+            trial_id:
+                Trial ID of a trial to add a parameter.
+            param_name:
+                Name of a parameter to add.
+            param_value_internal:
+                Internal representation of a value a parameter to add.
+            distribution:
+                Sampled distribution of a parameter to add.
+
+        Returns:
+            Return False when the parameter is already set to the trial.
```

Comment:
This only holds when each trial is updated from a single process.
We might explicitly add write-from-single-process as a requirement of this class.